# 03 — Modelagem

**Dimensão 5 da rúbrica — 20 pontos.**

Mínimo de **dois** classificadores distintos. Um único modelo zera 8 dos 20 pontos.


In [1]:
from pathlib import Path

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import HistGradientBoostingClassifier, RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import StratifiedKFold, cross_validate
%pip install -q imbalanced-learn
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline as ImbPipeline
from sklearn.metrics import confusion_matrix

RANDOM_STATE = 42

RAW = Path("..") / "data" / "raw" / "df.csv"
PROCESSED = Path("..") / "data" / "processed"
TARGET = "STATUS"

pd.set_option("display.max_columns", None)

Note: you may need to restart the kernel to use updated packages.


In [2]:
df = pd.read_csv(PROCESSED / "dataset_tratado.csv")

## 1. Split treino/teste

`stratify` preserva a proporção das classes nos dois conjuntos.


In [3]:
# Variáveis independentes
X = df.drop(columns=[TARGET])

# Variável Alvo
y = df[TARGET]

X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, stratify=y, random_state=RANDOM_STATE)

## 2. Modelos candidatos

Usar `Pipeline` evita vazamento: o scaler é ajustado só no fold de treino durante a validação cruzada.


In [ ]:
modelos = {
    # -------------------------------------------------------------------
    # MODELOS BASEADOS EM DISTÂNCIA/GRADIENTE: COM Scaler + SMOTE
    # Note que removemos o class_weight='balanced' para usar o balanceamento do SMOTE
    # -------------------------------------------------------------------
    "KNN": ImbPipeline([
        ("scaler", StandardScaler()),
        ("smote", SMOTE(random_state=RANDOM_STATE)),
        ("clf", KNeighborsClassifier(n_neighbors=5))
    ]),
    "Regressão Logística": ImbPipeline([
        ("scaler", StandardScaler()),
        ("smote", SMOTE(random_state=RANDOM_STATE)),
        ("clf", LogisticRegression(max_iter=1000, random_state=RANDOM_STATE))
    ]),
    "SVM": ImbPipeline([
        ("scaler", StandardScaler()),
        ("smote", SMOTE(random_state=RANDOM_STATE)),
        ("clf", SVC(random_state=RANDOM_STATE))
    ]),
    
    # -------------------------------------------------------------------
    # MODELOS BASEADOS EM ÁRVORES: SEM Scaler / SEM SMOTE
    # Usam o class_weight='balanced' nativo, que costuma performar melhor
    # -------------------------------------------------------------------
    "Floresta Aleatória": ImbPipeline([
        ("clf", RandomForestClassifier(n_estimators=100, random_state=RANDOM_STATE, n_jobs=-1, class_weight='balanced'))
    ]),
    "Árvore de Decisão": ImbPipeline([
        ("clf", DecisionTreeClassifier(random_state=RANDOM_STATE, class_weight='balanced'))
    ]),
    "Hist Gradient Boosting": ImbPipeline([
        ("clf", HistGradientBoostingClassifier(class_weight='balanced', random_state=RANDOM_STATE))
    ])
}

## 3. Validação cruzada


In [5]:
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
scoring = ['f1', 'accuracy', 'recall', 'precision']

resultados = []

for nome, modelo in modelos.items():
    scores = cross_validate(modelo, X_train, y_train, cv=cv, scoring=scoring, n_jobs=-1)
    resultados.append({
        "Modelo": nome,
        "F1-Score (Média)": scores['test_f1'].mean().round(2),
        "Accuracy (Média)": scores['test_accuracy'].mean().round(2),
        "Recall (Média)": scores['test_recall'].mean().round(2),
        "Precision (Média)": scores['test_precision'].mean().round(2)
    })

df_ranking = pd.DataFrame(resultados).sort_values(by="Modelo", ascending=False)
print(df_ranking.to_string(index=False))

                Modelo  F1-Score (Média)  Accuracy (Média)  Recall (Média)  Precision (Média)
     Árvore de Decisão              0.06              0.90            0.08               0.05
                   SVM              0.09              0.74            0.27               0.05
   Regressão Logística              0.10              0.59            0.49               0.05
                   KNN              0.07              0.79            0.17               0.04
Hist Gradient Boosting              0.09              0.85            0.16               0.06
    Floresta Aleatória              0.06              0.93            0.06               0.08


## 4. Matriz Confusão

In [6]:
from sklearn.metrics import ConfusionMatrixDisplay, confusion_matrix
cm = confusion_matrix(y_true, y_pred)
disp = ConfusionMatrixDisplay(confusion_matrix=cm)
disp.plot(cmap=plt.cm.Blues)
plt.show()

NameError: name 'y_true' is not defined

## 4. Comparação

Analisando a tabela de resultados, se verifica um cenário muito claro de forte desbalanceamento de classes (muito mais bons pagadores do que maus pagadores).

A resposta curta é: A **Regressão Logística** teve o melhor comportamento prático, mas todos os modelos estão com desempenho muito fraco para um problema real de concessão de crédito.

*Análise dos Modelos*
| Modelo | F1-Score | Accuracy | Recall| Precision| Avaliação no Contexto de Crédito|
|---|---|---|---|---|---|
|Regressão Logística|0.10|0.59|0.49|0.05|O "menos pior": Identifica a maior proporção de maus pagadores (Recall de 49%), embora a custa de muitos falsos alarmes.|
|SVM|0.09|0.74|0.27|0.05|Intermediário, captura apenas 27% dos inadimplentes.|
|Hist Gradient Boosting|0.09|0.85|0.16|0.06|Captura pouquíssimos inadimplentes (16%).|
|KNN|0.07|0.79|0.17|0.04|Baixa capacidade preditiva para a classe de risco.|
|Floresta Aleatória|0.06|0.93|0.06|0.08|A armadilha do desbalanceamento: Parece o "melhor" pela Acurácia (93%), mas é um dos piores — ignora 94% dos maus pagadores.|
|Árvore de Decisão|0.06|0.90|0.08|0.05|Praticamente inútil para prever a classe minoritária.|

Por que a Regressão Logística se destaca aqui?

Em análise de crédito, o custo de aprovar um mau pagador (inadimplência) é muito maior do que o custo de recusar um bom pagador. Portanto, a métrica mais crítica no início é o Recall (a capacidade de encontrar os maus pagadores no meio da base).

A Regressão Logística conseguiu capturar 49% dos maus pagadores (Recall = 0.49).

A Floresta Aleatória, apesar de ter a maior acurácia (93%), só pegou 6% dos maus pagadores (Recall = 0.06). Ela basicamente aprendeu a dizer "Aprovado" para quase todo mundo para acertar a maioria dos casos.

Por que os resultados gerais estão baixos (F1 ~ 0.06 a 0.10)?

Uma precisão de ~0.05 e F1-Score tão baixo indicam que a classe de maus pagadores é extremamente minoritária (provavelmente representa cerca de 1% a 5% da base total).

Para melhorar significativamente esses modelos antes da entrega final, aqui estão três ajustes prioritários:

Ajustar o Limiar de Decisão (Decision Threshold):Por padrão, os modelos usam $50\%$ de probabilidade para classificar como mau pagador. Em problemas de crédito muito desbalanceados, costuma-se baixar esse limiar para $10\%$ ou $20\%$ para tornar o modelo mais sensível à inadimplência.

Testar Métricas Adequadas para Desbalanceamento:Em vez de ordenar o ranking pelo F1-Score padrão, ordene pelo ROC-AUC ou PR-AUC (average_precision). Elas medem a capacidade do modelo de separar as duas classes independentemente do limite de probabilidade escolhido.

Engenharia de Recursos (Feature Engineering):Geralmente um desempenho tão baixo aponta que as variáveis atuais não possuem poder preditivo suficiente para separar bons e maus pagadores (ou faltam variáveis derivadas de comportamento recente).